File: JavaScript/009-special-pythagorean-triplet.js
/*
	A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,

	a² + b² = c²
	For example, 3² + 4² = 9 + 16 = 25 = 5².

	There exists exactly one Pythagorean triplet for which a + b + c = 1000. Find the product abc such that a + b + c = n.
*/

function specialPythagoreanTriplet(n) {
	// Prepare an array in case of multiple solutions
	const result = [];

	let a = 1;
	// Leave room for b and c, since a < b < c
	while (a < n - 2) {
		let b = a + 1;

		// Leave room for c which is larger than b
		while (b < n - 1) {
			let c = n - a - b; // a + b + c = n

			// Check if the numbers pass the test
			if(a**2 + b**2 === c**2) {
				let abc = a * b * c;
				// If they pass the test, add them to the results array, with their product
				result.push(`${a}, ${b}, ${c} => ${abc}`);
			}

			b++;
		}

		a++;
	}

	return result;
}
   
let triplets = specialPythagoreanTriplet(120);
console.log(triplets.join('\n')); // Show all results separated by line breaks

File: C++/009-special-pythagorean-triplet.cpp
/*
A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,

a2 + b2 = c2
For example, 32 + 42 = 9 + 16 = 25 = 52.

There exists exactly one Pythagorean triplet for which a + b + c = 1000.
Find the product abc.
*/

#include <iostream>
#include <cmath>
using namespace std;

void PythagorusTripletProduct (int sum) {

    int a,b,c;
    for (a = 1; a < sum/3; a++) {
        for (b = a + 1; b < sum/2; b++) {
            // a + b + c = sum
            c = sum - b - a;
            // check if a^2 + b^2 = c^2
            if (pow(a, 2) + pow(b, 2) == pow(c, 2)) {
                cout << "The Pythagorean triple is " << a << "," << b << "," << c << " and the sum is " << sum << "\n" ;
                cout << "The product abc is: " << a*b*c;
                break;
            }
        }
    }

}

int main () {
    
    PythagorusTripletProduct(1000);    
    return 0;
    
}